### PR TITLE
[TAS-553] ✨ Support uploading multiple files

### PR DESCRIPTION
**Pull request by AuroraHuang22**
截圖 2023-10-29 下午3 43 12
截圖 2023-10-29 下午3 44 17

### CODE REVIEW

File: components/IscnRegisterForm.vue [hunk 1]
```diff
@@ -135,9 +163,9 @@
             >{{ $t('IscnRegisterForm.label.numbersProtocol.details.link') }}</Link>
           </i18n>
         </CheckBox>
-      </FormField>
+      </FormField> -->
```
Comment (williamchong) [inline components/IscnRegisterForm.vue:166]: Please keep this checkbox, tick means upload all files to number, none means not

File: components/IscnRegisterForm.vue [hunk 2]
```diff
@@ -1194,6 +1216,14 @@ export default class IscnRegisterForm extends Vue {
     this.license = value
   }
 
+  formattedIpfs(ipfsHash: string) {
+    return this.$t('IscnRegisterForm.ipfs.link', { hash: ipfsHash })
+  }
+
+  formattedArweave(arweaveId: string) {
```
Comment (williamchong) [inline components/IscnRegisterForm.vue:1223]: method should be verb. e.g. `formatIpfs` and `formatArweave`

File: components/IscnRegisterForm.vue [hunk 3]
```diff
@@ -1069,6 +1088,9 @@ export default class IscnRegisterForm extends Vue {
   }
 
   async mounted() {
+    if (this.arweaveId) {
+      this.uploadArweaveIdList.push(this.arweaveId)
```
Comment (williamchong) [inline components/IscnRegisterForm.vue:1092]: Should reset `this.uploadArweaveIdList`? ie `this.uploadArweaveIdList = [this.arweaveId]`

File: components/IscnRegisterForm.vue [hunk 4]
```diff
@@ -1336,45 +1382,87 @@ export default class IscnRegisterForm extends Vue {
...
-    }
+    if (this.fileRecords.some(file => file.fileBlob)) {
+      try {
+        this.isUploadingArweave = true;
+        this.uploadStatus = 'uploading';
+        // eslint-disable-next-line no-restricted-syntax
+        for (const record of this.fileRecords) {
+          // eslint-disable-next-line no-await-in-loop
+          await this.submitToArweave(record);
+        }
```
Comment (williamchong) [inline components/IscnRegisterForm.vue:1393]: Should check and call `submitToNumbers`

File: components/IscnRegisterForm.vue [hunk 5]
```diff
@@ -1386,34 +1474,32 @@ export default class IscnRegisterForm extends Vue {
...
-    if (!this.fileBlob) return;
+    if (!tempRecord.fileBlob) return;
     this.isOpenSignDialog = true;
     this.onOpenKeplr();
-    let transactionHash = this.arweavePaymentTransactionHash
-    if (!transactionHash) {
-       transactionHash = await this.sendArweaveFeeTx();
+    tempRecord.transactionHash = this.sentArweaveTransactionHashes.get(tempRecord.ipfsHash)
+    if (!tempRecord.transactionHash) {
+     tempRecord.transactionHash = await this.sendArweaveFeeTx(tempRecord);
```
Comment (williamchong) [inline components/IscnRegisterForm.vue:1485]: Should also check if file is already uploaded and has a arweave Id

File: utils/cosmos/iscn/iscn.type.ts [hunk 6]
```diff
@@ -7,10 +7,10 @@ export interface ISCNRegisterPayload {
   url: string;
   exifInfo: any;
   license: string;
-  ipfsHash: string;
-  arweaveId: string;
+  ipfsHash: any;
+  arweaveId: any;
   numbersProtocolAssetId: string;
-  fileSHA256: string;
+  fileSHA256: any;
```
Comment (williamchong) [inline utils/cosmos/iscn/iscn.type.ts:13]: Please use `string[]` if we are using array only, `string | string[]` if either are possible

File: pages/new/index.vue [hunk 7]
```diff
@@ -157,55 +156,43 @@ export default class NewIndexPage extends Vue {
...
       if (url) {
         const { data, headers } = await this.$axios.get(url, { responseType: 'blob' })
-        this.fileBlob = data as Blob
-        this.isImage = headers['content-type'].startsWith('image')
-        this.fileType = headers['content-type']
-        this.fileSize = headers['content-length']
-        this.fileData = `data:${this.fileType};base64,${Buffer.from(await data.arrayBuffer(), 'binary').toString('base64')}`
+        this.urlFileRecords.push({
+        fileBlob:data as Blob,
+        ipfsHash:this.urlIpfsHash,
```
Comment (williamchong) [inline pages/new/index.vue:179]: Please fix indent and spaces

File: pages/new/index.vue [hunk 8]
```diff
@@ -157,55 +156,43 @@ export default class NewIndexPage extends Vue {
...
-      else if (this.ipfsHash) url = `https://ipfs.io/ipfs/${this.ipfsHash}`;
+      else if (this.urlIpfsHash) url = `https://ipfs.io/ipfs/${this.urlIpfsHash}`;
       if (url) {
         const { data, headers } = await this.$axios.get(url, { responseType: 'blob' })
-        this.fileBlob = data as Blob
-        this.isImage = headers['content-type'].startsWith('image')
-        this.fileType = headers['content-type']
-        this.fileSize = headers['content-length']
-        this.fileData = `data:${this.fileType};base64,${Buffer.from(await data.arrayBuffer(), 'binary').toString('base64')}`
+        this.urlFileRecords.push({
```
Comment (williamchong) [inline pages/new/index.vue:177]: Should reset `this.urlFileRecords`

File: components/IscnRegisterForm.vue [hunk 9]
```diff
@@ -1386,34 +1474,32 @@ export default class IscnRegisterForm extends Vue {
...
-        fileType: this.fileType as string,
-        txHash: transactionHash,
+        fileSize: tempRecord.fileBlob?.size || 0,
+        ipfsHash: tempRecord.ipfsHash,
+        fileType: tempRecord.fileType as string,
+        txHash: tempRecord.transactionHash,
       });
       if (arweaveId) {
-        this.uploadArweaveId = arweaveId
+        this.uploadArweaveIdList.push(arweaveId)
```
Comment (williamchong) [inline components/IscnRegisterForm.vue:1502]: Save the arweave Id back to `tempRecord` in case we will retry the whole batch

**Review by williamchong — CHANGES_REQUESTED** [review]

File: pages/nft/url/index.vue [hunk 10]
```diff
@@ -291,7 +291,7 @@ export default class FetchIndex extends Vue {
       description,
       tagsString: keywords,
       url: this.iscnData.url || this.url,
-      exifInfo: {},
+      exifInfo: [],
```
[the remark this answers is not in the document]
Comment (AuroraHuang22) [reply pages/nft/url/index.vue]: oops .. should be an object

File: components/IscnRegisterForm.vue [hunk 11]
```diff
@@ -826,7 +826,7 @@ export default class IscnRegisterForm extends Vue {
   authorName: string = ''
   authorUrl: string[] = []
   authorWalletAddress: string[] = []
-  sentArweaveTransactionHashes = new Map<string, string>()
+  sentArweaveTransactionHashes = new Map<string, any>()
```
Comment (williamchong) [inline components/IscnRegisterForm.vue:830]: if we already know it the `any` is a object with 2 optional key `transactionHash` and `arweaveId` we might as well define it

**Review by williamchong — APPROVED** [review]

**Pull request merged**